src/main/java/ch/njol/skript/expressions/ExprFormatNumber.java

```diff
@@ -0,0 +1,134 @@
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.PropertyExpression;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.VariableString;
+import ch.njol.skript.util.Getter;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.text.DecimalFormat;
+
+@Name("Formatted Number")
+@Description(
+	"Converts numbers to human-readable format. By default, '###,###' (e.g. '123,456,789') will be used. For reference, see this "
+		+ "<a href=\"https://docs.oracle.com/javase/7/docs/api/java/text/DecimalFormat.html\" target=\"_blank\">article</a>."
+)
+@Examples({
+	"command /formatnumber <number>:",
+		"\taliases: fn",
+		"\ttrigger:",
+			"\t\tsend \"Formatted: %formatted arg-1%\" to sender"
+})
+@Since("INSERT VERSION")
+public class ExprFormatNumber extends PropertyExpression<Number, String> {
+	
+	private static final DecimalFormat DEFAULT_FORMAT = new DecimalFormat("###,###");
+	
+	static {
+		Skript.registerExpression(ExprFormatNumber.class, String.class, ExpressionType.PROPERTY,
+			"%numbers% formatted [human-readable] [(with|as) %-string%]",
+			"[human-readable] formatted %numbers% [(with|as) %-string%]");
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private DecimalFormat format;
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<? extends String> customFormat;
+
+	@Override
+	@SuppressWarnings({"null", "unchecked"})
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		setExpr((Expression<? extends Number>) exprs[0]);
+		customFormat = (Expression<? extends String>) exprs[1];
+
+		boolean isSimpleString = customFormat instanceof VariableString && ((VariableString) customFormat).isSimple();
+		if (customFormat instanceof Literal || isSimpleString) {
+			String customFormatValue;
+			if (isSimpleString) {
+				customFormatValue = ((VariableString) customFormat).toString(null);
+			} else {
+				customFormatValue = ((Literal<String>) customFormat).getSingle();
+			}
```

Comment:
I think a comment should be placed above this literal condition. It looks really cursed but it's all to do parse time erroring of the format.